// ionic4-tooltips/esm5/tooltips.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TooltipBox } from './components/tooltip-box/tooltip-box.component';
import { TooltipController } from './controllers/tooltip.cotroller';
import { TooltipDirective } from './directives/tooltip.directive';
var TooltipsModule = /** @class */ (function () {
    function TooltipsModule() {
    }
    /**
     * @return {?}
     */
    TooltipsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: TooltipsModule,
            providers: [
                TooltipController
            ],
        };
    };
    TooltipsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        TooltipDirective,
                        TooltipBox
                    ],
                    entryComponents: [
                        TooltipBox
                    ],
                    exports: [
                        TooltipDirective
                    ],
                    imports: [
                        CommonModule
                    ]
                },] }
    ];
    return TooltipsModule;
}());
export { TooltipsModule };
